ably/realtime_presence.go

```diff
@@ -100,23 +130,60 @@ func (pres *RealtimePresence) syncWait() {
 	pres.syncMtx.Unlock()
 }
 
-func syncSerial(msg *protocolMessage) string {
-	if i := strings.IndexRune(msg.ChannelSerial, ':'); i != -1 {
-		return msg.ChannelSerial[i+1:]
+// RTP18
+func syncSerial(msg *protocolMessage) (noChannelSerial bool, syncSequenceId string, syncCursor string) {
+	if empty(msg.ChannelSerial) { // RTP18c
+		noChannelSerial = true
+		return
+	}
+	// RTP18a
+	serials := strings.Split(msg.ChannelSerial, ":")
+	syncSequenceId = serials[0]
+	if len(serials) > 1 {
+		syncCursor = serials[1]
 	}
-	return ""
+	return false, syncSequenceId, syncCursor
 }
 
-func (pres *RealtimePresence) onAttach(msg *protocolMessage) {
-	serial := syncSerial(msg)
+func (pres *RealtimePresence) enterMembers(internalMembers []*PresenceMessage) {
+	for _, member := range internalMembers {
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+		// RTP17g
+		err := pres.EnterClient(ctx, member.ClientID, member.Data)
+		// RTP17e
+		if err != nil {
+			pres.channel.log().Errorf("Error for internal member presence enter with id %v, clientId %v, err %v", member.ID, member.ClientID, err)
+			pres.channel.emitErrorUpdate(newError(91004, err), true)
+		}
+		cancel()
+	}
+}
+
+func (pres *RealtimePresence) onAttach(msg *protocolMessage, isAttachWithoutMessageLoss bool) {
 	pres.mtx.Lock()
 	defer pres.mtx.Unlock()
-	switch {
-	case msg.Flags.Has(flagHasPresence):
-		pres.syncStart(serial)
-	case pres.syncState == syncInitial:
-		pres.syncState = syncComplete
-		pres.syncMtx.Unlock()
+	// RTP1
+	if msg.Flags.Has(flagHasPresence) {
+		pres.syncStart()
+	} else {
+		pres.leaveMembers(pres.members) // RTP19a
+		if pres.syncState == syncInitial {
+			pres.syncState = syncComplete
+			pres.syncMtx.Unlock()
```

Comment:
This looks susceptible to deadlock, where does this get locked? Locks and unlocks should really always appear in pairs close to each other.

Reply:
This is also done because of `syncWait` method, you can check it's usage in comments
https://github.com/ably/ably-go/blob/e66ad6a1bfa0e0e48dccb7c1afd0c282b568d1c2/ably/realtime_presence.go#L93-L99

Reply:
Ok, that's really not a good use of a mutex, a mutex should be used to synchronise access to shared memory, not to synchronise clients waiting for the result of a network call (i.e. here waiting for a presence sync to complete). For example, what happens if the sync doesn't complete because the channel became detached? Does the mutex stay locked, and all calls to `syncWait` block until the program exits?

It's typical to use the closing of an empty struct channel to unblock multiple callers when some event has happened, and to select on it in a cancelable way:

```go
// a field on RealtimePresence
syncDone := make(chan struct{})

// when the sync has completed
close(p.syncDone)

// wait for the channel to close in syncWait
func (p *RealtimePresence) syncWait(ctx context.Context) error {
select {
case <-p.syncDone:
	return nil
case <-ctx.Done():
	return ctx.Err()
}
```

Reply:
sure, we can update

Reply:
Updated as a part of -> https://github.com/ably/ably-go/pull/621/commits/2740853cd88a58e6b0b6332904e61efcb3c4912c

Reply:
Ok thanks, can we get rid of `syncMtx` altogether?

Reply:
we did

Reply:
Yes sorry, I think I was looking at an outdated diff 👍